server/armeria-server/cats/src/main/scala/sttp/tapir/server/armeria/cats/ArmeriaCatsServerOptions.scala

```diff
@@ -64,17 +64,22 @@ object ArmeriaCatsServerOptions {
     })
 
   private def blocking[F[_], T](body: => T)(implicit F: Async[F]): F[T] = {
-    F.async_ { cb =>
-      CommonPools
-        .blockingTaskExecutor()
-        .execute(() => {
-          try {
-            cb(Right(body))
-          } catch {
-            case NonFatal(ex) =>
-              cb(Left(ex))
-          }
-        })
+    F.async { cb =>
+      F.delay {
+        val javaFuture = CommonPools
+          .blockingTaskExecutor()
+          .submit(new Runnable { 
+            override def run(): Unit = {
+              try {
+                cb(Right(body))
+              } catch {
+                case NonFatal(ex) =>
+                  cb(Left(ex))
+              }
+            }
+          })
+        Some(F.void(F.delay { javaFuture.cancel(true) }))
```

Comment:
shouldn't we wait for the future, but then wait for its result, to properly implement cancellation integration?

Reply:
You mean call the `cancel` but still block until the future completes? I think the finalizer should finish as soon as it finishes its logic. Take a look at this example from The `Async` documentation:

```scala
def fromCompletableFuture[A](fut: F[CompletableFuture[A]]): F[A] =
  flatMap(fut) { cf =>
    async[A] { cb =>
      delay {
        //Invoke the callback with the result of the completable future
        val stage = cf.handle[Unit] {
          case (a, null) => cb(Right(a))
          case (_, e) => cb(Left(e))
        }

        //Cancel the completable future if the fiber is canceled
        Some(void(delay(stage.cancel(false))))
      }
    }
  }
```

Reply:
I thought that the idea in cats-effect is for the finalizers to block until the effect finishes (after cancellation or not). But the example indeed suggests otherwise.

... and reading https://github.com/typelevel/cats-effect/releases/tag/v3.5.0 your code is correct :)